cirq/optimizers/three_qubit_decomposition.py

```diff
@@ -0,0 +1,276 @@
+        # we can represent the merge by phasing u2 with I ⊗ Z
+        u2 = u2 @ np.diag([1, -1, 1, -1])
+        cs_ops = cs_ops[:-1]
+
+    d_ud, ud_ops = _two_qubit_multiplexor_to_ops(q0, q1, q2, u1, u2, shift_left=True, atol=atol)
+
+    _, vdh_ops = _two_qubit_multiplexor_to_ops(
+        q0, q1, q2, v1h, v2h, shift_left=False, diagonal=d_ud, atol=atol
+    )
+
+    return list(cirq.Circuit(vdh_ops + cs_ops + ud_ops).all_operations())
+
+
+def _cs_to_ops(q0: ops.Qid, q1: ops.Qid, q2: ops.Qid, theta: np.ndarray) -> List[ops.Operation]:
+    """Converts theta angles based Cosine Sine matrix to operations.
+
+    Using the optimization as per Appendix A.1, it uses CZ gates instead of
+    CNOT gates and returns a circuit that skips the terminal CZ gate.
+
+    Args:
+        a, b, c: the 3 qubits in order
+        theta: theta returned from the Cosine Sine decomposition
+    Returns:
+         the operations
+    """
+    # Note: we are using *2 as the thetas are already half angles from the
+    # CSD decomposition, but cirq.ry takes full angles.
+    angles = _multiplexed_angles(theta * 2)
+    rys = [cirq.ry(angle).on(q0) for angle in angles]
+    ops = [
+        rys[0],
+        cirq.CZ(q1, q0),
+        rys[1],
+        cirq.CZ(q2, q0),
+        rys[2],
+        cirq.CZ(q1, q0),
+        rys[3],
+        cirq.CZ(q2, q0),
+    ]
+    return _optimize_multiplexed_angles_circuit(ops)
+
+
+def _optimize_multiplexed_angles_circuit(operations: Sequence[ops.Operation]):
+    """Removes two qubit gates that amount to identity.
+    Exploiting the specific multiplexed structure, this methods looks ahead
+    to find stripes of 3 or 4 consecutive CZ or CNOT gates and removes them.
+
+    Args:
+        operations: operations to be optimized
+    Returns:
+        the optimized operations
+    """
+    circuit = cirq.Circuit(operations)
+    cirq.optimizers.DropNegligible().optimize_circuit(circuit)
+    if np.allclose(circuit.unitary(), np.eye(8), atol=1e-14):
+        return cirq.Circuit([])
+
+    # the only way we can get identity here is if all four CZs are
+    # next to each other
+    def num_conseq_2qbit_gates(i):
+        j = i
+        while j < len(operations) and operations[j].gate.num_qubits() == 2:
+            j += 1
+        return j - i
+
+    operations = list(circuit.all_operations())
+
+    i = 0
+    while i < len(operations):
+        num_czs = num_conseq_2qbit_gates(i)
+        if num_czs == 4:
+            operations = operations[:1]
+            break
+        elif num_czs == 3:
+            operations = operations[:i] + [operations[i + 1]] + operations[i + 3 :]
+            break
+        else:
+            i += 1
+    return operations
+
+
+def _two_qubit_multiplexor_to_ops(
+    q0: ops.Qid,
+    q1: ops.Qid,
+    q2: ops.Qid,
+    u1: np.ndarray,
+    u2: np.ndarray,
+    shift_left: bool = True,
+    diagonal: np.ndarray = np.eye(4),
+    atol: float = 1e-8,
+) -> Tuple[np.ndarray, List[ops.Operation]]:
```

Comment:
Let's make it `Tuple[Optional[np.ndarray], List[ops.Operation]]` per other comment.